### PR TITLE
Reduce the number of type inlinings

### PR DESCRIPTION
**Pull request by ak3n**
This PR brings performance improvements in the inlining part:

1. Disables the beta-step for types
2. Adds `INLINE` pragma for `typeSubstTyNamesM`

Overall these changes reduce the compilation time one of the biggest marlowe contracts from 15s to 2s.

---

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs [hunk 1]
```diff
@@ -198,30 +198,28 @@ recursive and non-recursive data types with a single function.
...
 instance PLC.HasUniques (Term tyname name uni fun ann) => PLC.Rename (Program tyname name uni fun ann) where
     rename (Program ann term) = Program ann <$> PLC.rename term
 
 -- See Note [Renaming of constructors].
 -- | A wrapper around a function restoring some old context of the renamer.
-newtype Restorer m = Restorer
-    { unRestorer :: forall a. m a -> m a
+data Restorer = Restorer
+    { unRestorer :: forall a. PLC.ScopedRenameM a -> PLC.ScopedRenameM a
     }
```
Comment (ak3n) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs:210]: Not sure about this one. Can be improved somehow?

File: plutus-core/plutus-core.cabal [hunk 2]
```diff
@@ -34,7 +34,7 @@ common lang
                  -Wincomplete-uni-patterns -Wincomplete-record-updates
                  -Wredundant-constraints -Widentities
                  -Wnoncanonical-monoid-instances
-                 -Wunused-packages
+                 -Wunused-packages -fprof-auto
```
Comment (michaelpj) [inline plutus-core/plutus-core.cabal:37]: leftover prof options

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Beta.hs [hunk 3]
```diff
@@ -27,12 +27,13 @@ betaStep = \case
             bindings = binding :| []
         in
             Let a NonRec bindings body
-    TyInst a (TyAbs _ tyname kind body) typ ->
-        let tyVarDecl = TyVarDecl a tyname kind
-            tyBinding = TypeBind a tyVarDecl typ
-            bindings  = tyBinding :| []
-        in
-            Let a NonRec bindings body
+    -- This case is disabled as it introduces a lot of type inlining (determined from profiling)
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/Beta.hs:30]: Rather link to https://input-output.atlassian.net/browse/SCP-2570, the transformation is currently unsound! (will turn typechecking programs into non-typechecking programs)

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs:210]: I think it's okay? Why the change to data?

File: plutus-core/plutus-core/src/PlutusCore/Rename/Monad.hs [hunk 4]
```diff
@@ -30,10 +30,10 @@ import           Control.Lens
 import           Control.Monad.Reader
 
 -- | The monad the renamer runs in.
-newtype RenameT ren m a = RenameT
-    { unRenameT :: ReaderT ren m a
+newtype RenameM ren a = RenameM
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Rename/Monad.hs:33]: Comment about why this is as concrete as it is.

**Review by michaelpj — COMMENTED** [review]
Looks plausible. Would be good for Roman to have a quick look.

**Review by effectfully — COMMENTED** [review]
This PR does quite a lot of things and I agree with some of them and don't agree with others. Plus it's not clear how much of speedup comes from disabling beta-inlining for types and how much from the abstraction-butchering changes. The inliner does renaming in what is essentially a tight loop and so this is what makes the renamer look like a bottleneck. Now that you removed that (not sure if you removed all of that), do we think the renamer can be a bottleneck? LIke, can it take more time than type checking? I doubt that.

Changes I agree with:

1. Dropping `MonadReader` -- that's a clear improvement, I don't think we ever needed to be that abstract there (even though I liked how the code read)
2. Dropping `MonadQuote` -- also looks like an improvement, we can always use the concrete `Quote` and lift it into `MonadQuote` later.
3. sticking `INLINE` pragmas on functions (let's just stick them on every function in the internal module of the renamer and every relevant small function)

The change I really didn't like is hardcoding `Quote` inside `RenameM`. We don't need `Quote` for equality checking (even though we do what is renaming there) and so you essentially roll your own renaming machinery with the `ReaderT` monad and equivalents of the `lookupNameM` and `withRenamedName` combinators (the latter does not get duplicated though) for equality checking.

If we're going this route of having two different abstractions for renaming and equality checking, then we should decouple them fully, give a name to the monad that we do equality checking in, etc. But I'm not convinced that doing that is beneficial compared to the changes that I agreed with.

Overall, I believe before touching the renamer, it's worth fixing the inliner that does asymptotically a huge amount of work.

**Comment by michaelpj** [comment]
This is based on profiling work that Evgenii did, so we do have some evidence. Even after turning off the type inlining, renaming takes ~40% of the time, so it seemed worth optimizing it a bit. We should also definitely work on the inliner and ideally figure out a way to rename less, but making the renamer faster seems good anyway, since it gets called a lot in many places anyway.

**Comment by effectfully** [comment]
> Even after turning off the type inlining, renaming takes ~40% of the time

What are the numbers when the inliner is turned off completely? After all, the inliner calls the renamer in a place that this PR does not touch.

> but making the renamer faster seems good anyway, since it gets called a lot in many places anyway.

Not arguing with that at all. My suggestion amounts to keeping `RenameT` for it to be reused in the equality checking machinery and defining `type RenameM ren = RenameT ren Quote` (which is monomorphic again) and dropping `MonadReader` and `MonadQuote` (that the PR already does) and sticking `INLINE` on every relevant function that is not huge. Does that seem like it can ruin the speedup that this PR brings?

**Comment by michaelpj** [comment]
> Does that seem like it can ruin the speedup that this PR brings?

Not sure, we'd have to try! I don't think any of us quite have a handle on when the monad stack is monomorphic enough for GHC to inline things well...

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Beta.hs (hunk 3 above)
Comment (ak3n) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Beta.hs:30]: Yeah, it makes a dramatic improvement reducing the compilation time of the marlowe contract from 15s to 2s.

Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Beta.hs:30]: So there are a few problems here that this is covering up:
- This transformation is unsound
- The inlining for types is too eager and needs some heuristics, so overall we were inlining way too much
- We probably do too much renaming when we inline

This change happens to stop all that, but I want to tackle all of those issues separately. But even if we do all that I think we'll end up doing quite a lot of renaming, so I think we might as well make it faster.

**Comment by ak3n** [comment]
>Does that seem like it can ruin the speedup that this PR brings?

The nuisance here is that the renaming part is relevant only with this beta-step.

By turning off, the rest (concrete monad + INLINE pragmas) doesn't do anything useful much as they are simply called less often.

We may postpone these `Rename` refactorings until this beta step will be enabled but in a better way. And then profile again, because our current refactorings might be irrelevant in the future.

What do you think?

**Comment by effectfully** [comment]
@michaelpj 

> I don't think any of us quite have a handle on when the monad stack is monomorphic enough for GHC to inline things well...

I think the general rule is that if monomorphization happens in the same module, then it gets inlined, and otherwise it depends on `INLINE`/`INLINEABLE` pragmas etc. I propose to keep the general thing and monomorphize it in the same module, which should be fine, especially given the fact that pretty much all relevant functions are small. That's in addition to my "the renamer is not the problem here" point.

> By turning it off, the rest improvements (concrete monad + INLINE pragmas) don't do anything useful much as they are simply called less often.

That was my guess, yes.

> We may postpone these Rename refactorings until this beta step will be enabled but in a better way. And then profile again, because our current refactorings might be irrelevant in the future.

I think most of what is here is a definite improvement, so I suggest to rollback the removal of `RenameT` and addition of equality-specific `lookupNameM` and `withRenamedName` and just add a monomorphized `RenameM` to the same module where `RenameT` is defined and update every function mentioning `MonadQuote` to work over the monomorphic `RenameM`. And also stick `INLINE` pragmas on every relevant small function.

**Comment by michaelpj** [comment]
> I think most of what is here is a definite improvement, so I suggest to...

I'm also fine with this.

**Comment by ak3n** [comment]
🤔 

I'm having troubles with tests, and `test_scopingBad` in particular which overrides the `MonadReader` with newtypes. It's very annoying and makes me to question do we really want to use a concrete monad here now :D

**Comment by michaelpj** [comment]
If it's too painful we can shelve it. It's not so significant once we figured out the type inlining stuff, I mostly thought we might as well if you'd already done the work. I think pull out as much a is easy and still useful.

**Comment by ak3n** [comment]
Okay, thanks!

It feels like I did only a small part and there is a hidden refactoring that follows after these changes.

**Comment by michaelpj** [comment]
Ah, you will need to update some test output for the beta change, unfortunately.

**Comment by michaelpj** [comment]
`plutus-use-cases` too, I'm afraid.

**Comment by ak3n** [comment]
Ah, right, forgot about them, thanks!

**Comment by effectfully** [comment]
I looked into this and [here's](https://github.com/effectfully/plutus-prototype/commit/29997ff86396409be796e9856bedbde6633f89d1) how keeping the transformer would look like.

However that generalization of the renamer to work over any `MonadReader` has happened just recently in #3533 and precisely because we want to be able to supply different monads, so that we can test that a flawed monad (not handling scopes correctly, not allowing for any renaming to be done etc) gives you a misbehaved renamer. We could duplicate code or drop all the negative property tests for the renamer, but I would rather try using `SPECIALIZE` pragmas etc first, but then it doesn't seem like the renaming is a being a bottleneck, so I personally don't care.

**Pull request merged**